### PR TITLE
Use arm jit for armel in crossgen2, and build the matching bits on all architectures

### LINKED ISSUE #49347: armel build fail
armel build fails in System.Private.CoreLib crossgen2

How to
```
ROOTFS_DIR=../rootfs/armel ./build.sh --portablebuild false --arch armel --os Linux --clang10 --restore --build --configuration Release --cross --subset clr
```

Error Message
```
crossgen-corelib -> 
Generating native image of System.Private.CoreLib for Linux.armel.Release. Logging to /home/dheon/work/dotnet/jobs/net6/runtime/artifacts/log/CrossgenCoreLib_Linux__armel__Release.log
/home/dheon/work/dotnet/jobs/net6/runtime/dotnet.sh /home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/x64/crossgen2/crossgen2.dll -o:/home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/System.Private.CoreLib.dll -r:/home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/IL/*.dll --targetarch:armel -O /home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/IL/System.Private.CoreLib.dll --perfmap --perfmap-path:/home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/
EXEC : error : Unable to load shared library 'clrjit_unix_arm_x64' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libclrjit_unix_arm_x64: cannot open shared object file: No such file or directory [/home/dheon/work/dotnet/jobs/net6/runtime/src/coreclr/crossgen-corelib.proj]
  System.DllNotFoundException: Unable to load shared library 'clrjit_unix_arm_x64' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libclrjit_unix_arm_x64: cannot open shared object file: No such file or directory
     at System.Runtime.InteropServices.NativeLibrary.LoadByName(String libraryName, QCallAssembly callingAssembly, Boolean hasDllImportSearchPathFlag, UInt32 dllImportSearchPathFlag, Boolean throwOnError)
     at System.Runtime.InteropServices.NativeLibrary.LoadLibraryByName(String libraryName, Assembly assembly, Nullable`1 searchPath, Boolean throwOnError)
     at System.Runtime.InteropServices.NativeLibrary.Load(String libraryName, Assembly assembly, Nullable`1 searchPath)
     at Internal.JitInterface.JitConfigProvider.<>c__DisplayClass5_0.<Initialize>b__0(String libName, Assembly assembly, Nullable`1 searchPath)
     at System.Runtime.InteropServices.NativeLibrary.LoadLibraryCallbackStub(String libraryName, Assembly assembly, Boolean hasDllImportSearchPathFlags, UInt32 dllImportSearchPathFlags)
     at Internal.JitInterface.CorInfoImpl.jitStartup(IntPtr host)
     at Internal.JitInterface.CorInfoImpl.Startup()
     at Internal.JitInterface.JitConfigProvider.Initialize(TargetDetails target, IEnumerable`1 jitFlags, IEnumerable`1 parameters, String jitPath)
     at ILCompiler.ReadyToRunCodegenCompilationBuilder.ToCompilation()
     at ILCompiler.Program.Run(String[] args) in /home/dheon/work/dotnet/jobs/net6/runtime/src/coreclr/tools/aot/crossgen2/Program.cs:line 624
     at ILCompiler.Program.Main(String[] args) in /home/dheon/work/dotnet/jobs/net6/runtime/src/coreclr/tools/aot/crossgen2/Program.cs:line 791
/home/dheon/work/dotnet/jobs/net6/runtime/src/coreclr/crossgen-corelib.proj(107,5): error MSB3073: "/home/dheon/work/dotnet/jobs/net6/runtime/dotnet.sh /home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/x64/crossgen2/crossgen2.dll -o:/home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/System.Private.CoreLib.dll -r:/home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/IL/*.dll --targetarch:armel -O /home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/IL/System.Private.CoreLib.dll --perfmap --perfmap-path:/home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/" 명령이 종료되었습니다(코드: 1).

빌드하지 못했습니다.

EXEC : error : Unable to load shared library 'clrjit_unix_arm_x64' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libclrjit_unix_arm_x64: cannot open shared object file: No such file or directory [/home/dheon/work/dotnet/jobs/net6/runtime/src/coreclr/crossgen-corelib.proj]
/home/dheon/work/dotnet/jobs/net6/runtime/src/coreclr/crossgen-corelib.proj(107,5): error MSB3073: "/home/dheon/work/dotnet/jobs/net6/runtime/dotnet.sh /home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/x64/crossgen2/crossgen2.dll -o:/home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/System.Private.CoreLib.dll -r:/home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/IL/*.dll --targetarch:armel -O /home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/IL/System.Private.CoreLib.dll --perfmap --perfmap-path:/home/dheon/work/dotnet/jobs/net6/runtime/artifacts/bin/coreclr/Linux.armel.Release/" 명령이 종료되었습니다(코드: 1).
    경고 0개
    오류 2개
```

It try to load  libclrjit_unix_arm_x64.so instead of libclrjit_unix_armel_x64.so in crossgen2.
In my guess, it needs a PR (https://github.com/dotnet/runtime/pull/43706) to support armel in crossgen2. And updates targetspec string (below code) based on the PR.
https://github.com/dotnet/runtime/blob/59a2e12edd4563e28b011ab480cca2d98adaafb8/src/coreclr/tools/Common/JitInterface/JitConfigProvider.cs#L56
https://github.com/dotnet/runtime/blob/59a2e12edd4563e28b011ab480cca2d98adaafb8/src/coreclr/tools/Common/JitInterface/JitConfigProvider.cs#L136-L149

\+ In armel client, crossgen2 requires "libjitinterface_arm.so and libcjirjit_arm_arm.so" files instead of existing "libjitinterface_armel.so and libcjirjit_armel_armel.so".

### PR DESCRIPTION
**Pull request by davidwrighton**
Attempt to address use of crossgen2 on armel architectures. Fixes #49347

### CODE REVIEW

**Comment by davidwrighton** [comment]
@clamp03 I can't actually test this, as I don't know the docker image for doing an armel crossbuild. Can you provide a docker image that I can use here?

**Comment by davidwrighton** [comment]
The idea of the change is that the unix arm jit will be able to compile targetting soft_fp or normal hardware fp abi. And once we get rid of crossgen, we'll simply delete the armel jit we're currently building in favor of the configurable one.

**Comment by clamp03** [comment]
@davidwrighton Thank you for the update.

It fails to build armel with this patch in a restore stage.
```
/home/dheon/work/dotnet/jobs/net6/runtime/.dotnet/sdk/6.0.100-preview.2.21155.3/NuGet.RestoreEx.targets(19,5): error : An unexpected token "armel" was found at character position 26 in condition "'$(TargetArchitecture)=='armel'".  /home/d
heon/work/dotnet/jobs/net6/runtime/src/coreclr/tools/aot/crossgen2/crossgen2.csproj [/home/dheon/work/dotnet/jobs/net6/runtime/Build.proj
```

This is a Dockerfile that I put all instructions into. (without this patch).
```
FROM ubuntu:20.04

RUN apt-get update
RUN DEBIAN_FRONTEND="noninteractive" apt-get install -y build-essential git vim curl cmake make llvm-10 clang-10 lldb-10 liblldb-10-dev libunwind8 libunwind8-dev gettext libicu-dev liblttng-ust-dev libcurl4-openssl-dev libssl-dev libnuma-dev libkrb5-dev binutils-arm-linux-gnueabi zlib1g-dev libxml2-utils rpm2cpio cpio

WORKDIR /dotnet/
RUN git clone https://github.com/dotnet/runtime.git

WORKDIR /dotnet/runtime
RUN ROOTFS_DIR=/dotnet/rootfs/armel ./eng/common/cross/build-rootfs.sh armel tizen
RUN ROOTFS_DIR=/dotnet/rootfs/armel ./build.sh --portablebuild false --arch armel --os Linux --clang10 --restore --build --configuration Release --cross --subset clr+libs.native 
RUN ROOTFS_DIR=/dotnet/rootfs/armel ./build.sh --arch armel --os Linux --clang10 --restore --build --configuration Release --cross --subset libs.ref+libs.src+libs.pretest+libs.packages+libs.tests
```

Thank you.

**Comment by davidwrighton** [comment]
@clamp03, this looks to be working now although I see errors like the following "bash : error : setlocale: LC_ALL: cannot change locale (en_US.UTF-8) [/dotnet/runtime/src/libraries/Native/build-native.proj]" as it builds, but the failure to run crossgen2 seems to be fixed.

**Comment by davidwrighton** [comment]
Please let me know if the fix looks good in your scenario.

**Comment by davidwrighton** [comment]
@dotnet/crossgen-contrib

**Comment by clamp03** [comment]
@davidwrighton Thank you. And could you check two things?
- DEBUG BUILD FAIL: JIT_FLAG_SOFTFP_ABI is not defined.
https://github.com/dotnet/runtime/blob/eceb5c8951ca3e5902a6eecdd594f82916ff4762/src/coreclr/jit/compiler.cpp#L3255-L3257

- CROSSGEN2 FAIL in HOST(x64): ARM_SOFTFP & CONFIGURABLE_ARM_ABI are not defined in clrjit_arm_x64
https://github.com/dotnet/runtime/blob/eceb5c8951ca3e5902a6eecdd594f82916ff4762/src/coreclr/jit/CMakeLists.txt#L496-L499

File: src/coreclr/jit/CMakeLists.txt [hunk 1]
```diff
@@ -496,6 +496,9 @@ if (CLR_CMAKE_BUILD_SUBSET_ALLJITS AND NOT CLR_CROSS_COMPONENTS_BUILD)
 else()
   if (CLR_CMAKE_TARGET_UNIX)
     create_standalone_jit(TARGET clrjit_unix_${ARCH_TARGET_NAME}_${ARCH_HOST_NAME} OS unix ARCH ${ARCH_TARGET_NAME})
+    if (ARCH_TARGET_NAME STREQUAL arm)
```
Comment (clamp03) [inline src/coreclr/jit/CMakeLists.txt:499]: Isn't it better to use CLR_CMAKE_TARGET_ARCH_ARM?
```
if (CLR_CMAKE_TARGET_ARCH_ARM)
```

**Review by clamp03 — APPROVED** [review]
It works well in my scenario.
Thank you so much.

**Comment by davidwrighton** [comment]
@trylek could you review this? While @clamp03 has reviewed it, I need a reviewer with write access to also approve it.

@dotnet/crossgen-contrib

**Review by AntonLapounov — APPROVED** [review]

**Review by trylek — APPROVED** [review]
LGTM

**Pull request merged**